libs/payments/paypal/src/lib/paypal.manager.in.spec.ts

```diff
@@ -22,56 +24,82 @@ import {
+import { AmountExceedsPayPalCharLimitError } from './paypal.error';
 import {
   PaypalCustomerNotFoundError,
   PaypalCustomerMultipleRecordsError,
 } from './paypalCustomer/paypalCustomer.error';
-import { ResultPaypalCustomerFactory } from './paypalCustomer/paypalCustomer.factories';
+import {
+  BillingAgreementFactory,
+  ResultPaypalCustomerFactory,
+} from './paypalCustomer/paypalCustomer.factories';
 import { PaypalCustomerManager } from './paypalCustomer/paypalCustomer.manager';
 
 describe('PaypalManager', () => {
-  let kyselyDb: Kysely<DB>;
-  let paypalClient: PayPalClient;
   let paypalManager: PayPalManager;
-  let stripeClient: StripeClient;
-  let stripeConfig: StripeConfig;
+  let paypalClient: PayPalClient;
   let stripeManager: StripeManager;
+  let currencyManager: CurrencyManager;
   let paypalCustomerManager: PaypalCustomerManager;
 
   beforeAll(async () => {
-    kyselyDb = await testAccountDatabaseSetup([
-      'paypalCustomers',
-      'accountCustomers',
-    ]);
-
-    paypalClient = new PayPalClient({
-      sandbox: false,
-      user: faker.string.uuid(),
-      pwd: faker.string.uuid(),
-      signature: faker.string.uuid(),
-    });
-
-    stripeConfig = {
-      apiKey: faker.string.uuid(),
-      taxIds: { EUR: 'EU1234' },
-    };
-
-    stripeClient = new StripeClient({} as any);
-    stripeManager = new StripeManager(stripeClient, stripeConfig);
-    paypalCustomerManager = new PaypalCustomerManager(kyselyDb);
-
-    paypalManager = new PayPalManager(
-      kyselyDb,
-      paypalClient,
-      stripeManager,
-      paypalCustomerManager
+    const moduleRef = await Test.createTestingModule({
+      providers: [
+        PayPalManager,
+        PayPalClient,
+        StripeManager,
+        CurrencyManager,
+        PaypalCustomerManager,
+        {
+          provide: AccountDbProvider,
+          useValue: MockAccountDatabaseNestFactory,
+        },
+      ],
+    })
+      .overrideProvider(PayPalManager)
+      .useValue({
+        cancelBillingAgreement: jest.fn(),
+        getCustomerBillingAgreementId: jest.fn(),
+        getCheckoutToken: jest.fn(),
+        processNonZeroInvoice: jest.fn(),
+        processZeroInvoice: jest.fn(),
+        processInvoice: jest.fn(),
+        getPayPalAmountStringFromAmountInCents: jest.fn(),
+        createBillingAgreement: jest.fn(),
+        getBillingAgreement: jest.fn(),
+      })
+      .overrideProvider(PayPalClient)
+      .useValue({
+        baUpdate: jest.fn(),
+        setExpressCheckout: jest.fn(),
+      })
+      .overrideProvider(StripeManager)
+      .useValue({
+        getSubscriptions: jest.fn(),
+        finalizeInvoiceWithoutAutoAdvance: jest.fn(),
+        fetchActiveCustomer: jest.fn(),
+      })
+      .overrideProvider(CurrencyManager)
+      .useValue({
+        assertCurrencyCompatibleWithCountry: jest.fn(),
+      })
+      .overrideProvider(PaypalCustomerManager)
+      .useValue({
+        createPaypalCustomer: jest.fn(),
+      })
```

Comment:
I don't believe we've used this pattern anywhere else in our code-base. The other places where we've substituted values in for a provider, we do it as its done on line 54-55. Or we create the mock separately and use it, e.g.

```typescript
const mockPaypalManager = {
  cancelBillingAgreement: jest.fn(),
  ...
};
// Then in the providers
{
  provide: PayPalManager,
  useValue: mockPaypalManager
}
```

This also avoids having to do the moduleRef.get for each one, since the mock object is already defined in a reachable location.